xrootd/cmd/xrd-srv/handler.go

```diff
@@ -6,19 +6,31 @@ package main // import "go-hep.org/x/hep/xrootd/cmd/xrd-srv"
 
 import (
 	"fmt"
+	"io"
 	"io/ioutil"
+	"os"
 	"path"
+	"sync"
 
 	"go-hep.org/x/hep/xrootd/server"
 	"go-hep.org/x/hep/xrootd/xrdfs"
 	"go-hep.org/x/hep/xrootd/xrdproto"
 	"go-hep.org/x/hep/xrootd/xrdproto/dirlist"
+	"go-hep.org/x/hep/xrootd/xrdproto/open"
+	"go-hep.org/x/hep/xrootd/xrdproto/read"
+	"go-hep.org/x/hep/xrootd/xrdproto/xrdclose"
 )
 
 // handler implements server.Handler API by making request to the backing filesystem at basePath.
 type handler struct {
 	server.Handler
 	basePath string
+	sessions sync.Map
```

Comment:
is this needed in place of a simple `map[[16]byte]*session`?
the doc about `sync.Map` says:
```
    The Map type is optimized for two common use cases: (1) when the entry for a
    given key is only ever written once but read many times, as in caches that
    only grow, or (2) when multiple goroutines read, write, and overwrite
    entries for disjoint sets of keys. In these two cases, use of a Map may
    significantly reduce lock contention compared to a Go map paired with a
    separate Mutex or RWMutex.
```

I suspect you chose it because of `2)` ?
in that case, we should document a bit the thinking behind it.
(also, the output of `benchcmp` with and w/o `sync.Map` might help :P)

Reply:
ok, thanks for the explanation.
(perhaps put it somewhere either in the commit message or in the code?)

the magnitude of the alloc increase is somewhat surprising.
could you share the benchmark that produced it?

Reply:
Yep, [there](https://play.golang.org/p/G7I-lGdmZbH). I tried to simulate scenario when there is a lot of users that make a lot of `open`\`read`\`close` calls. So, each user adds a session, obtains it a lot, and, finally, deletes.

Reply:
hum... I got this:

```
$> go test -bench=. -benchmem -count=10 -memprofile=mem.out
goos: linux
goarch: amd64
BenchmarkSyncMap-4   	    2000	    828777 ns/op	   51572 B/op	     625 allocs/op
BenchmarkSyncMap-4   	    2000	    826259 ns/op	   51906 B/op	     625 allocs/op
BenchmarkSyncMap-4   	    2000	    815250 ns/op	   51912 B/op	     622 allocs/op
BenchmarkSyncMap-4   	    2000	    825395 ns/op	   51646 B/op	     624 allocs/op
BenchmarkSyncMap-4   	    2000	    822896 ns/op	   51854 B/op	     626 allocs/op
BenchmarkSyncMap-4   	    2000	    806667 ns/op	   50161 B/op	     639 allocs/op
BenchmarkSyncMap-4   	    2000	    807839 ns/op	   51957 B/op	     624 allocs/op
BenchmarkSyncMap-4   	    2000	    821685 ns/op	   52071 B/op	     622 allocs/op
BenchmarkSyncMap-4   	    2000	    818774 ns/op	   51988 B/op	     627 allocs/op
BenchmarkSyncMap-4   	    2000	    821352 ns/op	   51871 B/op	     623 allocs/op
BenchmarkMap-4       	    2000	    626359 ns/op	    1630 B/op	     101 allocs/op
BenchmarkMap-4       	    3000	    631235 ns/op	    1628 B/op	     101 allocs/op
BenchmarkMap-4       	    2000	    628663 ns/op	    1623 B/op	     101 allocs/op
BenchmarkMap-4       	    2000	    626625 ns/op	    1627 B/op	     101 allocs/op
BenchmarkMap-4       	    2000	    604494 ns/op	    1626 B/op	     101 allocs/op
BenchmarkMap-4       	    2000	    624580 ns/op	    1630 B/op	     101 allocs/op
BenchmarkMap-4       	    2000	    581111 ns/op	    1627 B/op	     101 allocs/op
BenchmarkMap-4       	    2000	    578524 ns/op	    1625 B/op	     101 allocs/op
BenchmarkMap-4       	    3000	    577159 ns/op	    1629 B/op	     101 allocs/op
BenchmarkMap-4       	    2000	    600275 ns/op	    1627 B/op	     101 allocs/op
PASS
ok  	_/home/binet/work/gsoc/src/go-hep.org/x/hep/xrootd	31.251s
```

Reply:
Interesting, I got:
```
BenchmarkSyncMap-4          2000            579450 ns/op           31769 B/op        613 allocs/op
BenchmarkSyncMap-4          3000            641949 ns/op           32527 B/op        608 allocs/op
BenchmarkSyncMap-4          3000            630656 ns/op           32521 B/op        603 allocs/op
BenchmarkSyncMap-4          3000            608373 ns/op           31787 B/op        600 allocs/op
BenchmarkSyncMap-4          2000            586932 ns/op           32447 B/op        593 allocs/op
BenchmarkSyncMap-4          3000            549863 ns/op           32479 B/op        587 allocs/op
BenchmarkSyncMap-4          3000            551857 ns/op           32407 B/op        585 allocs/op
BenchmarkSyncMap-4          3000            546538 ns/op           32466 B/op        584 allocs/op
BenchmarkSyncMap-4          3000            562828 ns/op           32508 B/op        587 allocs/op
BenchmarkSyncMap-4          3000            546538 ns/op           32379 B/op        586 allocs/op
BenchmarkMap-4              2000           1130478 ns/op            1625 B/op        101 allocs/op
BenchmarkMap-4              2000           1124992 ns/op            1626 B/op        101 allocs/op
BenchmarkMap-4              2000           1118508 ns/op            1627 B/op        101 allocs/op
BenchmarkMap-4              2000           1131474 ns/op            1626 B/op        101 allocs/op
BenchmarkMap-4              2000           1124493 ns/op            1629 B/op        101 allocs/op
BenchmarkMap-4              2000           1128981 ns/op            1626 B/op        101 allocs/op
BenchmarkMap-4              2000           1117014 ns/op            1629 B/op        101 allocs/op
BenchmarkMap-4              2000           1116016 ns/op            1627 B/op        101 allocs/op
BenchmarkMap-4              2000           1117512 ns/op            1627 B/op        101 allocs/op
BenchmarkMap-4              2000           1128982 ns/op            1626 B/op        101 allocs/op
```
under `Windows` with `go1.10.3` and with 4 virtual cores of  `i3-6100 CPU @ 2.30GHz`.

```
BenchmarkSyncMap-4          5000            319505 ns/op           34893 B/op        603 allocs/op
BenchmarkSyncMap-4          5000            319134 ns/op           34607 B/op        604 allocs/op
BenchmarkSyncMap-4          5000            318645 ns/op           34758 B/op        602 allocs/op
BenchmarkSyncMap-4          5000            319370 ns/op           34850 B/op        606 allocs/op
BenchmarkSyncMap-4          5000            320850 ns/op           34971 B/op        602 allocs/op
BenchmarkSyncMap-4          5000            320182 ns/op           34855 B/op        602 allocs/op
BenchmarkSyncMap-4          5000            319204 ns/op           34754 B/op        602 allocs/op
BenchmarkSyncMap-4          5000            320634 ns/op           34771 B/op        605 allocs/op
BenchmarkSyncMap-4          5000            318372 ns/op           34679 B/op        604 allocs/op
BenchmarkSyncMap-4          5000            318040 ns/op           34646 B/op        606 allocs/op
BenchmarkMap-4             10000            351543 ns/op            1623 B/op        101 allocs/op
BenchmarkMap-4              5000            343194 ns/op            1626 B/op        101 allocs/op
BenchmarkMap-4              5000            337057 ns/op            1624 B/op        101 allocs/op
BenchmarkMap-4              5000            346122 ns/op            1626 B/op        101 allocs/op
BenchmarkMap-4              5000            345659 ns/op            1626 B/op        101 allocs/op
BenchmarkMap-4              5000            346998 ns/op            1625 B/op        101 allocs/op
BenchmarkMap-4              5000            345495 ns/op            1623 B/op        101 allocs/op
BenchmarkMap-4              5000            350971 ns/op            1625 B/op        101 allocs/op
BenchmarkMap-4              5000            352238 ns/op            1624 B/op        101 allocs/op
BenchmarkMap-4              5000            347649 ns/op            1626 B/op        101 allocs/op
```
under `Linux` with `go1.10.3` and with 4 physical cores of `i5-7600 CPU @ 3.50GHz`.

and (really suprising):
```
BenchmarkSyncMap-4          2000            608910 ns/op           35745 B/op        748 allocs/op
BenchmarkSyncMap-4          3000            595478 ns/op           35278 B/op        749 allocs/op
BenchmarkSyncMap-4          3000            592310 ns/op           35309 B/op        750 allocs/op
BenchmarkSyncMap-4          2000            590734 ns/op           35150 B/op        750 allocs/op
BenchmarkSyncMap-4          3000            603761 ns/op           35272 B/op        751 allocs/op
BenchmarkSyncMap-4          3000            589368 ns/op           35295 B/op        749 allocs/op
BenchmarkSyncMap-4          3000            589488 ns/op           35350 B/op        749 allocs/op
BenchmarkSyncMap-4          3000            597075 ns/op           34924 B/op        751 allocs/op
BenchmarkSyncMap-4          3000            590481 ns/op           35227 B/op        750 allocs/op
BenchmarkSyncMap-4          2000            590959 ns/op           35194 B/op        749 allocs/op
BenchmarkMap-4              2000            645056 ns/op            1628 B/op        101 allocs/op
BenchmarkMap-4              2000            627111 ns/op            1624 B/op        101 allocs/op
BenchmarkMap-4              2000            623458 ns/op            1626 B/op        101 allocs/op
BenchmarkMap-4              2000            626699 ns/op            1622 B/op        101 allocs/op
BenchmarkMap-4              2000            644903 ns/op            1633 B/op        101 allocs/op
BenchmarkMap-4              2000            630497 ns/op            1629 B/op        101 allocs/op
BenchmarkMap-4              2000            709978 ns/op            1624 B/op        101 allocs/op
BenchmarkMap-4              2000            658578 ns/op            1627 B/op        101 allocs/op
BenchmarkMap-4              2000            661421 ns/op            1629 B/op        101 allocs/op
BenchmarkMap-4              2000            643000 ns/op            1626 B/op        101 allocs/op
```
under [`WSL`](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux) with `go1.10` and with 4 virtual cores of  `i3-6100 CPU @ 2.30GHz`.

Looks like under `Linux` `map` works better that under `Windows`.

Probably it's better to go with `map` instead of `sync.Map` then due to less memory usage?
What do you think?

P.S. I tried to change `cnt` from `100` to `500` and got same results in terms of relative difference.

Reply:
I was on Linux with Go 1.11b2.

Yeah, let's go with 'map' for now.